File: VoucherManagementBE/src/main/java/com/example/vouchermanagementbe/module/voucher/domain/repository/IVoucherUsageRepository.java
package com.example.vouchermanagementbe.module.voucher.domain.repository;

import com.example.vouchermanagementbe.core.PageResponse;
import com.example.vouchermanagementbe.module.voucher.domain.entity.VoucherUsage;

public interface IVoucherUsageRepository {
    VoucherUsage save(VoucherUsage usage);
    PageResponse<VoucherUsage> findAll(int pageNo, int pageSize);
}

File: VoucherManagementFE/src/core/domain/repositories/IVoucherUsageRepository.ts
import { VoucherUsage } from "../entities/VoucherUsage";
import { ApiResponse, PaginatedResponse } from "../entities/ApiResponse";

/**
 * IVoucherUsageRepository - Repository Interface for VoucherUsage Entity
 * Defines contract for voucher usage history data access operations
 * Implementation should be provided by infrastructure layer
 */
export interface IVoucherUsageRepository {
  /**
   * Get all voucher usages (paginated)
   */
  getAll(page?: number, size?: number): Promise<ApiResponse<PaginatedResponse<VoucherUsage>>>;

  /**
   * Get voucher usages by user ID (paginated)
   */
  getByUserId(userId: number, page?: number, size?: number): Promise<ApiResponse<PaginatedResponse<VoucherUsage>>>;

  /**
   * Get voucher usages by voucher ID (paginated)
   */
  getByVoucherId(voucherId: number, page?: number, size?: number): Promise<ApiResponse<PaginatedResponse<VoucherUsage>>>;

  /**
   * Create new voucher usage record
   */
  create(voucherUsage: VoucherUsage): Promise<ApiResponse<VoucherUsage>>;
}
